sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreCheckpointFormatV2Suite.scala

```diff
@@ -211,6 +218,317 @@ class RocksDBStateStoreCheckpointFormatV2Suite extends StreamTest
     }
   }
 
+  val changelogEnabled =
+    "spark.sql.streaming.stateStore.rocksdb.changelogCheckpointing.enabled" -> "true"
+  val changelogDisabled =
+    "spark.sql.streaming.stateStore.rocksdb.changelogCheckpointing.enabled" -> "false"
+  val ckptv2 = SQLConf.STATE_STORE_CHECKPOINT_FORMAT_VERSION.key -> "2"
+
+  case class checkpointV2CompatibilityTest(
+    configsForFirstRun: Seq[(String, String)],
+    configsForSecondRun: Seq[(String, String)]
+  ) {
+    override def toString: String = "first run - " +
+      s"changeLogEnabled: ${configsForFirstRun(0)._2}, ckpt ver: ${configsForFirstRun(1)._2}" +
+      s" - second run - changeLogEnabled: ${configsForSecondRun(0)._2}, " +
+      s"ckpt ver: ${configsForSecondRun(1)._2}"
+  }
+
+  val testConfigSetups = Seq(
+    // Enable and disable changelog under ckpt v2
+    checkpointV2CompatibilityTest(Seq(changelogEnabled, ckptv2), Seq(changelogEnabled, ckptv2)),
+    checkpointV2CompatibilityTest(Seq(changelogDisabled, ckptv2), Seq(changelogDisabled, ckptv2))
+  )
+
+  testConfigSetups.foreach { testConfig =>
+    testWithRocksDBStateStore("checkpointFormatVersion2 Backward Compatibility - simple agg - " +
```

Comment:
Consider whether we really need to test every stateful operator, or we just test 1-2 and it should be enough. I think backward compatibility should be mostly inside RocksDB state store, and as long as we guarantee that proper checkpointID and conf are passed into RocksDB state store (I hope other tests already covered it), we should be fine. This is different from the first checkpointV2 PR, as there we need some validation to make sure that checkpointID is passed back and force for all those stateful operators correctly. Here I'm not sure.

Reply:
Thanks for the comment! In general I believe more tests don't do harm. But if you believe some specific test should just be removed I'd be willing to comply

Reply:
More tests mean more code to maintain, CI runs longer and more overhead dealing with flaky tests. We should not put more tests for no reason. For those compatibility cases, I would think perhaps keeping one is enough. You can pick whatever query. If you really would like to be solid, you can pick such cases:
1. one from either aggregation or Dedup
2. one for flatMapGroupsWithState
3. one for stream stream join
4. one for TransformWithState
5. one for chained operator.

1-4 are to cover different places of calling `StateStore.get()`. 5 is to cover more than one stateful operators. But they are already covered in the initial PR where we would make sure the correct checkpointID is passed in. I'm not sure the low chance that correct checkpointID is passed in but still backward compatibility is broken is worth a new unit test. Perhaps it can cover a regression in the future where those call paths diverge, but I doubt the need of maintaining too many tests just for that. Maybe we should cover things like this in a randomized/fuzzy test in the future when we develop it. I would vote for simplicity and don't have those coverage for now.

Comparing to this, I think it is more important to cover the case of multiple switches between changelog on or off, etc.

Reply:
> multiple switches between changelog on or off, etc.

This is already added in this PR

Reply:
Removed the tests